### ISSUE TITLE
Define the properties of Padding

### ISSUE THREAD

**Issue by Havvy** [rust-lang/reference#882]
We don't have a good explanation of the properties of padding. Currently we talk about it in two places:

1. Type layout; It's pointed out that a type can have alignment padding that adds to the size.
2. Added in #852, it's pointed out that it can be set as uninitialized.

I think padding should probably be described in the type layout chapter.

My question though, other than being an unnameable psuedo-field that can theoretically have any value, does it have any other properties? E.g., can they be repurposed by enums to do niche optimizations (making it unsafe to set the value)? Other restrictions?

**Comment by RalfJung** [comment]
> My question though, other than being an unnameable psuedo-field that can theoretically have any value, does it have any other properties?

Yes, a very crucial one: padding content is lost on "typed copies" (assignment, function argument/return passing). The content of the padding is not guaranteed to be preserved by such actions.

That's also why it can *not* be used for niche optimizations.

**Comment by Havvy** [comment]
Is there a difference between a "typed copy" and a move?

**Comment by RalfJung** [comment]
I am not sure what you mean by "move". ;) Both the MIR `Operand::Copy` and `Operand::Move` are "typed copies". There might be a difference between them, that is discussed at https://github.com/rust-lang/unsafe-code-guidelines/issues/188.
